src/table/use-sticky-columns.tsx

```diff
@@ -0,0 +1,228 @@
+import { useVisualRefresh } from '../internal/hooks/use-visual-mode';
+import { useIntersectionObserver } from '../internal/hooks/use-intersection-observer';
+import { TableProps } from './interfaces';
+interface CellWidths {
+  start: number[];
+  end: number[];
+}
+interface StickyStyles {
+  left?: string;
+  right?: string;
+  paddingLeft?: string;
+}
+
+interface StickyColumnParams {
+  containerWidth: number | null;
+  hasSelection: boolean;
+  isWrapperScrollable: boolean;
+  stickyColumns?: TableProps.StickyColumns;
+  visibleColumnsLength: number;
+  tableRefObject: React.RefObject<HTMLTableElement>;
+  wrapperRefObject: React.RefObject<HTMLDivElement>;
+}
+
+export interface GetStickyColumn {
+  isStickyLeft: boolean;
+  isStickyRight: boolean;
+  isLastStickyLeft: boolean;
+  isLastStickyRight: boolean;
+  stickyStyles: StickyStyles;
+}
+
+const getPadding = (element: HTMLElement | null, side: 'Left' | 'Right') =>
+  element ? Number(getComputedStyle(element)[`padding${side}`].slice(0, -2)) : 0;
+
+// We allow the table to have a minimum of 148px of available space besides the sum of the widths of the sticky columns
+const MINIMUM_SCROLLABLE_SPACE = 148;
+
+export const useStickyState = ({
+  tableRefObject,
+  wrapperRefObject,
+}: {
+  tableRefObject: React.RefObject<HTMLTableElement>;
+  wrapperRefObject: React.RefObject<HTMLDivElement>;
+}) => {
+  const [isStuckToTheLeft, setIsStuckToTheLeft] = useState(false);
+  const [isStuckToTheRight, setIsStuckToTheRight] = useState(false);
+
+  const intersectionObserverOptions = { threshold: [0, 1], rootMargin: '-2px' };
+  const { ref: leftSentinelRef, isIntersecting: leftSentinelIntersecting } =
+    useIntersectionObserver(intersectionObserverOptions);
+  const { ref: rightSentinelRef, isIntersecting: rightSentinelIntersecting } =
+    useIntersectionObserver(intersectionObserverOptions);
+  useEffect(() => {
+    setIsStuckToTheLeft(!leftSentinelIntersecting);
+    setIsStuckToTheRight(!rightSentinelIntersecting);
+  }, [leftSentinelIntersecting, rightSentinelIntersecting]);
+
+  useEffect(() => {
+    const wrapper = wrapperRefObject.current;
+    const table = tableRefObject.current;
+    if (!wrapper) {
+      return;
+    }
+    // Check the scrolling position of the table wrapper to set the initial "stuck" state
+    const right = wrapper.scrollLeft < wrapper.scrollWidth - wrapper.clientWidth - getPadding(table, 'Right');
+    const left = wrapper.scrollLeft > getPadding(table, 'Left');
+    setIsStuckToTheLeft(left);
+    setIsStuckToTheRight(right);
+  }, [wrapperRefObject, tableRefObject]);
+
+  return { isStuckToTheLeft, isStuckToTheRight, leftSentinelRef, rightSentinelRef };
+};
+
+export const useCellWidths = (tableCellRefs: Array<React.RefObject<HTMLTableCellElement>>) => {
+  const [cellWidths, setCellWidths] = useState<CellWidths>({ start: [], end: [] });
+
+  const updateCellWidths = useCallback(() => {
```

Comment:
What is the purpose of this function?

Reply:
Is it to inform the hook when the sticky columns are resized by the user so that the width can be set?

If so, why this separate state is needed, cannot we use the columnDefinitions?

Reply:
Shouldn't updateCellWidths be called when data changes? I would imagine that when e.g. the user change pages the widths are updated by the browser when not explicitly set.

Reply:
This function keeps track of the computed offsets of each of the cells. It produces an array for the values for the first columns offset (left-to-right) and one for the last columns (right-to-left).
You can image it produces an array like this: `[0, 120, 400, 600, 900, 1200]`. This allows us to set the appropriate `left` or `right` styles for the sticky columns depending on their indexes.
I updated the function name to better reflect its functionality and added some comments. 
The column definitions can't be used because we need the actual computed offset values to guarantee the styles work as expected.